src/metabase/driver/sql_jdbc/sync/describe_database.clj

```diff
@@ -0,0 +1,122 @@
+  (:require [clojure.java.jdbc :as jdbc]
+            [clojure.string :as str]
+            [metabase
+             [driver :as driver]
+             [util :as u]]
+            [metabase.driver.sql-jdbc
+             [connection :as sql-jdbc.conn]
+             [execute :as sql-jdbc.execute]]
+            [metabase.driver.sql-jdbc.sync
+             [common :as common]
+             [interface :as i]])
+  (:import [java.sql Connection DatabaseMetaData]))
+
+(defmethod i/excluded-schemas :sql-jdbc [_] nil)
+
+(defn- all-schemas [^DatabaseMetaData metadata]
+  {:pre [(instance? DatabaseMetaData metadata)]}
+  (reify clojure.lang.IReduceInit
+    (reduce [_ rf init]
+      (with-open [rs (.getSchemas metadata)]
+        (transduce
+         (map :table_schem)
+         rf
+         init
+         (jdbc/reducible-result-set rs {}))))))
+
+(defn- syncable-schemas
+  [driver metadata]
+  (eduction (remove (set (i/excluded-schemas driver)))
+            (all-schemas metadata)))
+
+(defn- execute-select-probe-query
+  "Execute the simple SELECT query defined above. The main goal here is to check whether we're able to execute a SELECT
+  query against the Table in question -- we don't care about the results themselves -- so the query and the logic
+  around executing it should be as simple as possible. We need to highly optimize this logic because it's executed for
+  every Table on every sync."
+  [driver ^Connection conn [sql & params]]
+  {:pre [(string? sql)]}
+  (with-open [stmt (common/prepare-statement driver conn sql params)]
+    ;; attempting to execute the SQL statement will throw an Exception if we don't have permissions; otherwise it will
+    ;; truthy wheter or not it returns a ResultSet, but we can ignore that since we have enough info to proceed at
+    ;; this point.
+    (.execute stmt)))
+
+(defmethod i/have-select-privilege? :sql-jdbc
+  [driver conn table-schema table-name]
+  ;; Query completes = we have SELECT privileges
+  ;; Query throws some sort of no permissions exception = no SELECT privileges
+  (u/ignore-exceptions
+    (execute-select-probe-query driver conn
+                                (common/simple-select-probe-query driver table-schema table-name))
+    true))
+
+(defn- db-tables
+  "Fetch a JDBC Metadata ResultSet of tables in the DB, optionally limited to ones belonging to a given
+  schema. Returns a reducible sequence of results."
+  [driver ^DatabaseMetaData metadata ^String schema-or-nil ^String db-name-or-nil]
+  (reify
+    clojure.lang.IReduceInit
+    (reduce [_ rf init]
+      ;; tablePattern "%" = match all tables
+      (with-open [rs (.getTables metadata db-name-or-nil (driver/escape-entity-name-for-metadata driver schema-or-nil) "%"
+                                 (into-array String ["TABLE" "VIEW" "FOREIGN TABLE" "MATERIALIZED VIEW" "EXTERNAL TABLE"]))]
+        (reduce
+         rf
+         init
+         (eduction (map #(select-keys % [:table_name :remarks :table_schem]))
```

Comment:
Why do we need an eduction here, doesn't it get consumed immediately by reduce?

Reply:
it does, but to apply the transducer we'd have to use `transduce` instead of `reduce` and that doesn't work exactly the same since it calls the  completing arity of `rf` (meaning it could end up getting called more than once depending on what the caller is doing -- I don't think you're supposed to `transduce` as the implementation of `IReduceInit`. `reduce` over an `eduction` avoids doing weird things like calling the completing arity of `rf`

Reply:
Actually I guess that's not really true, I guess we could apply the transducer to the `rf` directly and still use `reduce`

Reply:
Yes, I think you can do something like: 
`(reduce ((map ...) rf) init ...)`

Reply:
that's exactly what `eduction` does i believe.